### PR TITLE
Claw

### PR DESCRIPTION
**Pull request by KinetekEnergy**
# Claw
- review pls

### CODE REVIEW

File: Team 3749 2023/src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -27,11 +32,18 @@ public RobotContainer() {
...
-  private void configureDefaultCommands() {}
+  private void configureDefaultCommands() {
+  }
 
-  private void configureButtonBindings() {}
+  private void configureButtonBindings() {
+    pilot.a().whileTrue(new ClawIntakeCommand(claw)); // while a is held, intake
+    pilot.b().whileTrue(new ClawOuttakeCommand(claw)); // while b is held, outtake
+
+    // note that isHeld() doesn't work
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/RobotContainer.java]: this is an unecessary comment

File: Team 3749 2023/src/main/java/frc/robot/utils/Constants.java [hunk 2]
```diff
@@ -11,5 +13,35 @@ public static final class Base {
...
+
+        public static final int speed = 0; // number for the speed
+        public static final int stop = 0; // speed 0 = motors stop
+
+        //PID values
+        public static final double claw_kP = 0.1;
+        public static final double claw_kI = 0;
+        public static final double claw_kD = 0;
+
+        public static final int setpoint = 0;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/utils/Constants.java]: you should call this setpoint_velocity

File: Team 3749 2023/src/main/java/frc/robot/commands/Clawauto.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package frc.robot.commands;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/Clawauto.java]: delete this class if its not ready to  be tested

File: Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        double encoder_AVG = (left_encoder.getVelocity() + right_encoder.getVelocity())/2;
+        return(encoder_AVG);
+    }
+        /**
+     * set speed for motor
+     * @param setpoint
+     * This parmeter is to tell the PID calculator what the ideal speed is, and how to get there.
+     * @return 
+     */
+
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java]: delete empty lines

File: Team 3749 2023/src/main/java/frc/robot/commands/ClawIntakeCommand.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+/***
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/ClawIntakeCommand.java]: this should be above the class definition not above the imports

File: Team 3749 2023/src/main/java/frc/robot/commands/ClawIntakeCommand.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) FIRST and other WPILib contributors.
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/ClawIntakeCommand.java]: remove these copyright comments

File: Team 3749 2023/src/main/java/frc/robot/commands/ClawOuttakeCommand.java [hunk 7]
```diff
@@ -1,27 +1,31 @@
 // Copyright (c) FIRST and other WPILib contributors.
 // Open Source Software; you can modify and/or share it under the terms of
 // the WPILib BSD license file in the root directory of this project.
+/***
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/ClawOuttakeCommand.java]: move these to above the class not abov ethe imports

File: Team 3749 2023/src/main/java/frc/robot/commands/ClawOuttakeCommand.java [hunk 8]
```diff
@@ -1,27 +1,31 @@
 // Copyright (c) FIRST and other WPILib contributors.
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/ClawOuttakeCommand.java]: delete these copyright coments

File: Team 3749 2023/src/main/java/frc/robot/commands/ClawSimulationCommand.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
+// Copyright (c) FIRST and other WPILib contributors.
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/ClawSimulationCommand.java]: delete these copyright comments

File: Team 3749 2023/src/main/java/frc/robot/commands/ClawSimulationCommand.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+/***
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/ClawSimulationCommand.java]: move author comments to above the class, not to above the imports

File: Team 3749 2023/src/main/java/frc/robot/commands/ClawSimulationCommand.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
+public class ClawSimulationCommand extends CommandBase {
+    @SuppressWarnings({ "PMD.UnusedPrivateField", "PMD.SingularField" })
+
+    ClawSim clawSim = new ClawSim();
+
+    // left and right side of the claw (the motor)
+    private CANSparkMax right_motor = new CANSparkMax(Constants.Claw.right_side, MotorType.kBrushless);
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/ClawSimulationCommand.java]: why are these in the command?

File: Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
+/***
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java]: move to above the class, not the imports

File: Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    private final RelativeEncoder right_encoder = right_motor.getEncoder();
+    private final RelativeEncoder left_encoder = left_motor.getEncoder();
+    // Initializes the base subsystem
+    public Claw() {
+        right_motor.setInverted(true); // invert the motor to not break it
+
+        right_motor.setIdleMode(IdleMode.kBrake); // set neo to be braked when not active
+        left_motor.setIdleMode(IdleMode.kBrake); // set neo to be braked when not active
+
+        // not sure what this is: Constants.Base.speed.set(new Double(16.90));
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java]: delete unecessary comments

File: Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+import com.revrobotics.RelativeEncoder;
+
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.utils.Constants;
+
+public class Claw extends SubsystemBase {
+    // Creates a PIDController with gains kP, kI, and kD
+    PIDController claw_PID = new PIDController(Constants.Claw.claw_kP, Constants.Claw.claw_kI, Constants.Claw.claw_kD);
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java]: encapsulate this object

File: Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+import com.revrobotics.CANSparkMax.IdleMode;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+import com.revrobotics.RelativeEncoder;
+
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.utils.Constants;
+
+public class Claw extends SubsystemBase {
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java:26]: your variables should be private and final

File: Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    public Claw() {
+        right_motor.setInverted(true); // invert the motor to not break it
+
+        right_motor.setIdleMode(IdleMode.kBrake); // set neo to be braked when not active
+        left_motor.setIdleMode(IdleMode.kBrake); // set neo to be braked when not active
+
+        // not sure what this is: Constants.Base.speed.set(new Double(16.90));
+    }
+    
+    // averages the Encoder velocities from both left and right encoders.(gives error if method is not set to return double)
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java]: add javadoc comments

File: Team 3749 2023/src/main/java/frc/robot/subsystems/ClawSim.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.CANSparkMax.IdleMode;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+import com.revrobotics.RelativeEncoder;
+
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.utils.Constants;
+
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/ClawSim.java]: move author comments here

File: Team 3749 2023/src/main/java/frc/robot/utils/ColorSensor.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
+/***
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/utils/ColorSensor.java]: delete this class if its not being used, move your author comments

File: Team 3749 2023/src/main/java/frc/robot/utils/ColorSensor.Java [hunk 19]
```diff
@@ -0,0 +1,58 @@
+package frc.robot.utils;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/utils/ColorSensor.Java]: why are there 2 ColorSensor.java?

File: Team 3749 2023/src/main/java/frc/robot/subsystems/ClawSim.java [hunk 20]
```diff
@@ -0,0 +1,70 @@
+/***
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/ClawSim.java]: move author comments

**Review by rohinsood — CHANGES_REQUESTED** [review]
The commands, subsystems, and files need some formatting and standardization changes, other than that it looks good.

File: Team 3749 2023/src/main/java/frc/robot/commands/ClawSimulationCommand.java [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+import com.revrobotics.RelativeEncoder;
+
+import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+
+/***
+ * @author Zaddy Harkirat
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/ClawSimulationCommand.java]: mane what is this delete

File: Team 3749 2023/src/main/java/frc/robot/commands/ClawSimulationCommand.java [hunk 22]
```diff
@@ -0,0 +1,63 @@
+package frc.robot.commands;
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/commands/ClawSimulationCommand.java]: delete unused files

File: Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+    // left and right side of the claw (the motor)
+    private final CANSparkMax right_motor = new CANSparkMax(Constants.Claw.right_side, MotorType.kBrushless);
+    private final CANSparkMax left_motor = new CANSparkMax(Constants.Claw.left_side, MotorType.kBrushless);
+
+    // motor controller group for both sides
+    private final MotorControllerGroup clawMotors = new MotorControllerGroup(left_motor, right_motor);
+
+    // relative encoder
+    private final RelativeEncoder right_encoder = right_motor.getEncoder();
+    private final RelativeEncoder left_encoder = left_motor.getEncoder();
```
Comment (rohinsood) [inline Team 3749 2023/src/main/java/frc/robot/subsystems/Claw.java]: add an empty line below this

**Review by rohinsood — APPROVED** [review]
I went and fixed your code code code regarding formatting conventions and new commands, check my commits

**Review by rjawesome — APPROVED** [review]

**Pull request merged**